File: Core/evaluation_harness.py
"""
Core/evaluation_harness.py — RAG Evaluation Harness for ZEN_RAG.

Phase 3.3: Compute standard IR and NLP metrics for RAG quality assessment.

Metrics:
  - NDCG@k  (Normalized Discounted Cumulative Gain)
  - MRR     (Mean Reciprocal Rank)
  - Hit@k   (fraction of queries with at least one relevant doc in top-k)
  - BLEU    (BiLingual Evaluation Understudy, requires nltk)
  - ROUGE-L (Longest Common Subsequence recall, requires rouge-score)
  - Precision@k, Recall@k

Usage:
    harness = EvaluationHarness(rag=my_rag, llm=my_llm)
    results = harness.run(test_cases=[
        {"query": "What is X?", "expected": "X is ...", "relevant_docs": ["url1"]},
    ])
    harness.print_report(results)
    harness.save_report(results, "eval_results.json")
"""

import json
import logging
import math
import time
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, List

logger = logging.getLogger(__name__)


# =============================================================================
# Metric Implementations (pure Python, no heavy deps)
# =============================================================================


def _dcg(relevances: List[float], k: int) -> float:
    """Discounted Cumulative Gain at k."""
    return sum(rel / math.log2(i + 2) for i, rel in enumerate(relevances[:k]))


def _ndcg(retrieved: List[float], ideal: List[float], k: int) -> float:
    """NDCG@k. relevances[i] = relevance score of i-th retrieved doc."""
    idcg = _dcg(sorted(ideal, reverse=True), k)
    if idcg == 0:
        return 0.0
    return _dcg(retrieved, k) / idcg


def _mrr(retrieved: List[float]) -> float:
    """Mean Reciprocal Rank (expects relevances list; returns 1/rank for first relevant)."""
    for rank, rel in enumerate(retrieved, 1):
        if rel > 0:
            return 1.0 / rank
    return 0.0


def _hit_at_k(retrieved: List[float], k: int) -> float:
    """1 if any relevant doc in top-k, else 0."""
    return 1.0 if any(r > 0 for r in retrieved[:k]) else 0.0


def _precision_at_k(retrieved: List[float], k: int) -> float:
    """Fraction of top-k retrieved docs that are relevant."""
    if k == 0:
        return 0.0
    return sum(1 for r in retrieved[:k] if r > 0) / k


def _recall_at_k(retrieved: List[float], n_relevant: int, k: int) -> float:
    """Fraction of all relevant docs found in top-k."""
    if n_relevant == 0:
        return 0.0
    return sum(1 for r in retrieved[:k] if r > 0) / n_relevant


def _bleu_score(reference: str, hypothesis: str) -> float:
    """Compute sentence BLEU-1 (unigram precision). No external deps."""
    ref_tokens = reference.lower().split()
    hyp_tokens = hypothesis.lower().split()
    if not hyp_tokens or not ref_tokens:
        return 0.0
    ref_set = set(ref_tokens)
    matches = sum(1 for t in hyp_tokens if t in ref_set)
    return matches / len(hyp_tokens)


def _rouge_l(reference: str, hypothesis: str) -> float:
    """ROUGE-L F1 via LCS (pure Python)."""
    ref = reference.lower().split()
    hyp = hypothesis.lower().split()
    if not ref or not hyp:
        return 0.0

    # LCS dynamic programming
    m, n = len(ref), len(hyp)
    dp = [[0] * (n + 1) for _ in range(m + 1)]
    for i in range(1, m + 1):
        for j in range(1, n + 1):
            dp[i][j] = dp[i - 1][j - 1] + 1 if ref[i - 1] == hyp[j - 1] else max(dp[i - 1][j], dp[i][j - 1])
    lcs = dp[m][n]

    precision = lcs / n
    recall = lcs / m
    if precision + recall == 0:
        return 0.0
    return 2 * precision * recall / (precision + recall)


# =============================================================================
# EvaluationHarness
# =============================================================================


class EvaluationHarness:
    """
    End-to-end RAG evaluation harness.

    Supports:
      - Retrieval metrics: NDCG@k, MRR, Hit@k, Precision@k, Recall@k
      - Generation metrics: BLEU, ROUGE-L
      - Latency tracking
      - JSON report export
    """

    def __init__(
        self,
        rag: Any,
        llm: Any = None,
        k: int = 5,
        output_dir: str = "eval_results",
    ):
        """
        Args:
            rag: LocalRAG instance.
            llm: Optional LLM for answer generation evaluation.
            k: Top-k for retrieval metrics.
            output_dir: Directory for saved reports.
        """
        self.rag = rag
        self.llm = llm
        self.k = k
        self.output_dir = Path(output_dir)
        self.output_dir.mkdir(parents=True, exist_ok=True)

    def run(
        self,
        test_cases: List[Dict],
        verbose: bool = True,
    ) -> Dict:
        """
        Run evaluation over a list of test cases.

        Each test case dict:
          - query (str): The question
          - expected (str): Reference answer for BLEU/ROUGE
          - relevant_docs (List[str]): URLs/titles of relevant documents
          - relevance_scores (Dict[str, float]): Optional graded relevance (0-1)

        Returns:
            Aggregated metrics dict + per-sample results.
        """
        per_sample = []
        t0_total = time.time()

        for i, tc in enumerate(test_cases):
            query = tc.get("query", "")
            expected = tc.get("expected", "")
            relevant_docs = set(tc.get("relevant_docs", []))
            rel_scores_map = tc.get("relevance_scores", {})

            if not query:
                continue

            if verbose:
                logger.info(f"[Eval] Sample {i + 1}/{len(test_cases)}: {query[:60]}...")

            # Retrieve
            t0 = time.time()
            try:
                retrieved = self.rag.hybrid_search(query, k=self.k, rerank=True)
            except Exception as e:
                logger.warning(f"[Eval] Search failed for query '{query}': {e}")
                retrieved = []
            latency = time.time() - t0

            # Compute relevance labels
            relevances = self._compute_relevances(retrieved, relevant_docs, rel_scores_map)
            ideal = sorted(relevances, reverse=True)

            # Retrieval metrics
            sample_metrics = {
                "query": query,
                "latency_s": round(latency, 3),
                "n_retrieved": len(retrieved),
                f"ndcg@{self.k}": round(_ndcg(relevances, ideal, self.k), 4),
                "mrr": round(_mrr(relevances), 4),
                f"hit@{self.k}": _hit_at_k(relevances, self.k),
                f"precision@{self.k}": round(_precision_at_k(relevances, self.k), 4),
                f"recall@{self.k}": round(_recall_at_k(relevances, len(relevant_docs), self.k), 4),
                "bleu": 0.0,
                "rouge_l": 0.0,
                "generated_answer": "",
            }

            # Generation metrics (if LLM available and expected answer given)
            if self.llm and expected:
                generated = self._generate_answer(query, retrieved)
                sample_metrics["generated_answer"] = generated
                if generated:
                    sample_metrics["bleu"] = round(_bleu_score(expected, generated), 4)
                    sample_metrics["rouge_l"] = round(_rouge_l(expected, generated), 4)

            per_sample.append(sample_metrics)

        total_time = time.time() - t0_total

        # Aggregate
        agg = self._aggregate(per_sample, total_time)
        return {"summary": agg, "samples": per_sample}

    def _compute_relevances(
        self,
        retrieved: List[Dict],
        relevant_docs: set,
        rel_scores_map: Dict[str, float],
    ) -> List[float]:
        """Compute relevance score for each retrieved chunk."""
        relevances = []
        for chunk in retrieved:
            url = chunk.get("url", "") or ""
            title = chunk.get("title", "") or ""
            combined = url + " " + title

            # Graded relevance if provided
            if rel_scores_map:
                score = max(
                    rel_scores_map.get(url, 0.0),
                    rel_scores_map.get(title, 0.0),
                )
                relevances.append(score)
            else:
                # Binary: relevant if url/title matches any relevant_doc
                is_rel = any(rd in combined or combined in rd for rd in relevant_docs if rd)
                relevances.append(1.0 if is_rel else 0.0)

        return relevances

    def _generate_answer(self, query: str, context_chunks: List[Dict]) -> str:
        """Generate answer using LLM for BLEU/ROUGE evaluation."""
        try:
            context = "\n\n".join(c.get("text", "")[:500] for c in context_chunks[:5])
            prompt = f"Answer concisely based on context:\n\nContext:\n{context}\n\nQuestion: {query}\n\nAnswer:"
            if hasattr(self.llm, "query_sync"):
                return self.llm.query_sync(prompt, max_tokens=200).strip()
            elif hasattr(self.llm, "generate"):
                return self.llm.generate(prompt).strip()
        except Exception as e:
            logger.warning(f"[Eval] LLM generation failed: {e}")
        return ""

    def _aggregate(self, per_sample: List[Dict], total_time: float) -> Dict:
        """Compute mean metrics across all samples."""
        if not per_sample:
            return {}

        metric_keys = [k for k in per_sample[0] if isinstance(per_sample[0][k], (int, float))]
        agg = {
            "n_samples": len(per_sample),
            "total_time_s": round(total_time, 2),
            "avg_latency_s": round(sum(s["latency_s"] for s in per_sample) / len(per_sample), 3),
        }
        for key in metric_keys:
            values = [s[key] for s in per_sample if key in s]
            if values:
                agg[f"mean_{key}"] = round(sum(values) / len(values), 4)

        return agg

    def print_report(self, results: Dict):
        """Print formatted evaluation report."""
        summary = results.get("summary", {})
        print("\n" + "=" * 60)
        print("ZEN_RAG EVALUATION REPORT")
        print("=" * 60)
        print(f"Samples evaluated : {summary.get('n_samples', 0)}")
        print(f"Total time        : {summary.get('total_time_s', 0):.1f}s")
        print(f"Avg latency       : {summary.get('avg_latency_s', 0):.3f}s")
        print("-" * 60)
        for key, val in summary.items():
            if key.startswith("mean_") and "latency" not in key:
                metric_name = key.replace("mean_", "").upper()
                print(f"{metric_name:<20}: {val:.4f}")
        print("=" * 60)

    def save_report(self, results: Dict, filename: str = None) -> Path:
        """Save evaluation results to JSON."""
        if not filename:
            ts = datetime.now().strftime("%Y%m%d_%H%M%S")
            filename = f"eval_{ts}.json"
        out_path = self.output_dir / filename
        with open(out_path, "w", encoding="utf-8") as f:
            json.dump(results, f, indent=2, ensure_ascii=False)
        logger.info(f"[Eval] Report saved to {out_path}")
        return out_path

File: rust_output/src/evaluation_harness.rs
/// Core/evaluation_harness::py — RAG Evaluation Harness for ZEN_RAG.
/// 
/// Phase 3.3: Compute standard IR and NLP metrics for RAG quality assessment.
/// 
/// Metrics:
/// - NDCG@k  (Normalized Discounted Cumulative Gain)
/// - MRR     (Mean Reciprocal Rank)
/// - Hit@k   (fraction of queries with at least one relevant doc in top-k)
/// - BLEU    (BiLingual Evaluation Understudy, requires nltk)
/// - ROUGE-L (Longest Common Subsequence recall, requires rouge-score)
/// - Precision@k, Recall@k
/// 
/// Usage:
/// harness = EvaluationHarness(rag=my_rag, llm=my_llm)
/// results = harness.run(test_cases=[
/// {"query": "What is X?", "expected": "X is ...", "relevant_docs": ["url1"]},
/// ])
/// harness.print_report(results)
/// harness.save_report(results, "eval_results.json")

use anyhow::{Result, Context};
use std::collections::HashMap;
use std::collections::HashSet;
use std::fs::File;
use std::io::{self, Read, Write};
use std::path::PathBuf;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

/// End-to-end RAG evaluation harness.
/// 
/// Supports:
/// - Retrieval metrics: NDCG@k, MRR, Hit@k, Precision@k, Recall@k
/// - Generation metrics: BLEU, ROUGE-L
/// - Latency tracking
/// - JSON report export
#[derive(Debug, Clone)]
pub struct EvaluationHarness {
    pub rag: String,
    pub llm: String,
    pub k: String,
    pub output_dir: PathBuf,
}

impl EvaluationHarness {
    /// Args:
    /// rag: LocalRAG instance.
    /// llm: Optional LLM for answer generation evaluation.
    /// k: Top-k for retrieval metrics.
    /// output_dir: Directory for saved reports.
    pub fn new(rag: Box<dyn std::any::Any>, llm: Box<dyn std::any::Any>, k: i64, output_dir: String) -> Self {
        Self {
            rag,
            llm,
            k,
            output_dir: PathBuf::from(output_dir),
        }
    }
    /// Run evaluation over a list of test cases.
    /// 
    /// Each test case dict:
    /// - query (str): The question
    /// - expected (str): Reference answer for BLEU/ROUGE
    /// - relevant_docs (List[str]): URLs/titles of relevant documents
    /// - relevance_scores (Dict[str, float]): Optional graded relevance (0-1)
    /// 
    /// Returns:
    /// Aggregated metrics dict + per-sample results.
    pub fn run(&mut self, test_cases: Vec<HashMap>, verbose: bool) -> Result<HashMap> {
        // Run evaluation over a list of test cases.
        // 
        // Each test case dict:
        // - query (str): The question
        // - expected (str): Reference answer for BLEU/ROUGE
        // - relevant_docs (List[str]): URLs/titles of relevant documents
        // - relevance_scores (Dict[str, float]): Optional graded relevance (0-1)
        // 
        // Returns:
        // Aggregated metrics dict + per-sample results.
        let mut per_sample = vec![];
        let mut t0_total = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64();
        for (i, tc) in test_cases.iter().enumerate().iter() {
            let mut query = tc.get(&"query".to_string()).cloned().unwrap_or("".to_string());
            let mut expected = tc.get(&"expected".to_string()).cloned().unwrap_or("".to_string());
            let mut relevant_docs = tc.get(&"relevant_docs".to_string()).cloned().unwrap_or(vec![]).into_iter().collect::<HashSet<_>>();
            let mut rel_scores_map = tc.get(&"relevance_scores".to_string()).cloned().unwrap_or(HashMap::new());
            if !query {
                continue;
            }
            if verbose {
                logger.info(format!("[Eval] Sample {}/{}: {}...", (i + 1), test_cases.len(), query[..60]));
            }
            let mut t0 = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64();
            // try:
            {
                let mut retrieved = self.rag.hybrid_search(query, /* k= */ self.k, /* rerank= */ true);
            }
            // except Exception as e:
            let mut latency = (std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64() - t0);
            let mut relevances = self._compute_relevances(retrieved, relevant_docs, rel_scores_map);
            let mut ideal = { let mut v = relevances.clone(); v.sort(); v };
            let mut sample_metrics = HashMap::from([("query".to_string(), query), ("latency_s".to_string(), ((latency as f64) * 10f64.powi(3)).round() / 10f64.powi(3)), ("n_retrieved".to_string(), retrieved.len()), (format!("ndcg@{}", self.k), ((_ndcg(relevances, ideal, self.k) as f64) * 10f64.powi(4)).round() / 10f64.powi(4)), ("mrr".to_string(), ((_mrr(relevances) as f64) * 10f64.powi(4)).round() / 10f64.powi(4)), (format!("hit@{}", self.k), _hit_at_k(relevances, self.k)), (format!("precision@{}", self.k), ((_precision_at_k(relevances, self.k) as f64) * 10f64.powi(4)).round() / 10f64.powi(4)), (format!("recall@{}", self.k), ((_recall_at_k(relevances, relevant_docs.len(), self.k) as f64) * 10f64.powi(4)).round() / 10f64.powi(4)), ("bleu".to_string(), 0.0_f64), ("rouge_l".to_string(), 0.0_f64), ("generated_answer".to_string(), "".to_string())]);
            if (self.llm && expected) {
                let mut generated = self._generate_answer(query, retrieved);
                sample_metrics["generated_answer".to_string()] = generated;
                if generated {
                    sample_metrics["bleu".to_string()] = ((_bleu_score(expected, generated) as f64) * 10f64.powi(4)).round() / 10f64.powi(4);
                    sample_metrics["rouge_l".to_string()] = ((_rouge_l(expected, generated) as f64) * 10f64.powi(4)).round() / 10f64.powi(4);
                }
            }
            per_sample.push(sample_metrics);
        }
        let mut total_time = (std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64() - t0_total);
        let mut agg = self._aggregate(per_sample, total_time);
        Ok(HashMap::from([("summary".to_string(), agg), ("samples".to_string(), per_sample)]))
    }
    /// Compute relevance score for each retrieved chunk.
    pub fn _compute_relevances(&self, retrieved: Vec<HashMap>, relevant_docs: HashSet<String>, rel_scores_map: HashMap<String, f64>) -> Vec<f64> {
        // Compute relevance score for each retrieved chunk.
        let mut relevances = vec![];
        for chunk in retrieved.iter() {
            let mut url = (chunk.get(&"url".to_string()).cloned().unwrap_or("".to_string()) || "".to_string());
            let mut title = (chunk.get(&"title".to_string()).cloned().unwrap_or("".to_string()) || "".to_string());
            let mut combined = ((url + " ".to_string()) + title);
            if rel_scores_map {
                let mut score = rel_scores_map.get(&url).cloned().unwrap_or(0.0_f64).max(rel_scores_map.get(&title).cloned().unwrap_or(0.0_f64));
                relevances.push(score);
            } else {
                let mut is_rel = relevant_docs.iter().filter(|rd| rd).map(|rd| (combined.contains(&rd) || rd.contains(&combined))).collect::<Vec<_>>().iter().any(|v| *v);
                relevances.push(if is_rel { 1.0_f64 } else { 0.0_f64 });
            }
        }
        relevances
    }
    /// Generate answer using LLM for BLEU/ROUGE evaluation.
    pub fn _generate_answer(&mut self, query: String, context_chunks: Vec<HashMap>) -> Result<String> {
        // Generate answer using LLM for BLEU/ROUGE evaluation.
        // try:
        {
            let mut context = context_chunks[..5].iter().map(|c| c.get(&"text".to_string()).cloned().unwrap_or("".to_string())[..500]).collect::<Vec<_>>().join(&"\n\n".to_string());
            let mut prompt = format!("Answer concisely based on context:\n\nContext:\n{}\n\nQuestion: {}\n\nAnswer:", context, query);
            if /* hasattr(self.llm, "query_sync".to_string()) */ true {
                self.llm.query_sync(prompt, /* max_tokens= */ 200).trim().to_string()
            } else if /* hasattr(self.llm, "generate".to_string()) */ true {
                self.llm.generate(prompt).trim().to_string()
            }
        }
        // except Exception as e:
        Ok("".to_string())
    }
    /// Compute mean metrics across all samples.
    pub fn _aggregate(&self, per_sample: Vec<HashMap>, total_time: f64) -> HashMap {
        // Compute mean metrics across all samples.
        if !per_sample {
            HashMap::new()
        }
        let mut metric_keys = per_sample[0].iter().filter(|k| /* /* isinstance(per_sample[0][&k], (int, float) */) */ true).map(|k| k).collect::<Vec<_>>();
        let mut agg = HashMap::from([("n_samples".to_string(), per_sample.len()), ("total_time_s".to_string(), ((total_time as f64) * 10f64.powi(2)).round() / 10f64.powi(2)), ("avg_latency_s".to_string(), (((per_sample.iter().map(|s| s["latency_s".to_string()]).collect::<Vec<_>>().iter().sum::<i64>() / per_sample.len()) as f64) * 10f64.powi(3)).round() / 10f64.powi(3))]);
        for key in metric_keys.iter() {
            let mut values = per_sample.iter().filter(|s| s.contains(&key)).map(|s| s[&key]).collect::<Vec<_>>();
            if values {
                agg[format!("mean_{}", key)] = (((values.iter().sum::<i64>() / values.len()) as f64) * 10f64.powi(4)).round() / 10f64.powi(4);
            }
        }
        agg
    }
    /// Print formatted evaluation report.
    pub fn print_report(&self, results: HashMap<String, serde_json::Value>) -> () {
        // Print formatted evaluation report.
        let mut summary = results.get(&"summary".to_string()).cloned().unwrap_or(HashMap::new());
        println!("{}", ("\n".to_string() + ("=".to_string() * 60)));
        println!("{}", "ZEN_RAG EVALUATION REPORT".to_string());
        println!("{}", ("=".to_string() * 60));
        println!("Samples evaluated : {}", summary.get(&"n_samples".to_string()).cloned().unwrap_or(0));
        println!("Total time        : {:.1}s", summary.get(&"total_time_s".to_string()).cloned().unwrap_or(0));
        println!("Avg latency       : {:.3}s", summary.get(&"avg_latency_s".to_string()).cloned().unwrap_or(0));
        println!("{}", ("-".to_string() * 60));
        for (key, val) in summary.iter().iter() {
            if (key.starts_with(&*"mean_".to_string()) && !key.contains(&"latency".to_string())) {
                let mut metric_name = key.replace(&*"mean_".to_string(), &*"".to_string()).to_uppercase();
                println!("{:<20}: {:.4}", metric_name, val);
            }
        }
        println!("{}", ("=".to_string() * 60));
    }
    /// Save evaluation results to JSON.
    pub fn save_report(&mut self, results: HashMap<String, serde_json::Value>, filename: String) -> Result<PathBuf> {
        // Save evaluation results to JSON.
        if !filename {
            let mut ts = datetime::now().strftime("%Y%m%d_%H%M%S".to_string());
            let mut filename = format!("eval_{}.json", ts);
        }
        let mut out_path = (self.output_dir / filename);
        let mut f = File::create(out_path)?;
        {
            json::dump(results, f, /* indent= */ 2, /* ensure_ascii= */ false);
        }
        logger.info(format!("[Eval] Report saved to {}", out_path));
        Ok(out_path)
    }
}

/// Discounted Cumulative Gain at k.
pub fn _dcg(relevances: Vec<f64>, k: i64) -> f64 {
    // Discounted Cumulative Gain at k.
    relevances[..k].iter().enumerate().iter().map(|(i, rel)| (rel / math::log2((i + 2)))).collect::<Vec<_>>().iter().sum::<i64>()
}

/// NDCG@k. relevances[i] = relevance score of i-th retrieved doc.
pub fn _ndcg(retrieved: Vec<f64>, ideal: Vec<f64>, k: i64) -> f64 {
    // NDCG@k. relevances[i] = relevance score of i-th retrieved doc.
    let mut idcg = _dcg({ let mut v = ideal.clone(); v.sort(); v }, k);
    if idcg == 0 {
        0.0_f64
    }
    (_dcg(retrieved, k) / idcg)
}

/// Mean Reciprocal Rank (expects relevances list; returns 1/rank for first relevant).
pub fn _mrr(retrieved: Vec<f64>) -> f64 {
    // Mean Reciprocal Rank (expects relevances list; returns 1/rank for first relevant).
    for (rank, rel) in retrieved.iter().enumerate().iter() {
        if rel > 0 {
            (1.0_f64 / rank)
        }
    }
    0.0_f64
}

/// 1 if any relevant doc in top-k, else 0.
pub fn _hit_at_k(retrieved: Vec<f64>, k: i64) -> f64 {
    // 1 if any relevant doc in top-k, else 0.
    if retrieved[..k].iter().map(|r| r > 0).collect::<Vec<_>>().iter().any(|v| *v) { 1.0_f64 } else { 0.0_f64 }
}

/// Fraction of top-k retrieved docs that are relevant.
pub fn _precision_at_k(retrieved: Vec<f64>, k: i64) -> f64 {
    // Fraction of top-k retrieved docs that are relevant.
    if k == 0 {
        0.0_f64
    }
    (retrieved[..k].iter().filter(|r| r > 0).map(|r| 1).collect::<Vec<_>>().iter().sum::<i64>() / k)
}

/// Fraction of all relevant docs found in top-k.
pub fn _recall_at_k(retrieved: Vec<f64>, n_relevant: i64, k: i64) -> f64 {
    // Fraction of all relevant docs found in top-k.
    if n_relevant == 0 {
        0.0_f64
    }
    (retrieved[..k].iter().filter(|r| r > 0).map(|r| 1).collect::<Vec<_>>().iter().sum::<i64>() / n_relevant)
}

/// Compute sentence BLEU-1 (unigram precision). No external deps.
pub fn _bleu_score(reference: String, hypothesis: String) -> f64 {
    // Compute sentence BLEU-1 (unigram precision). No external deps.
    let mut ref_tokens = reference.to_lowercase().split_whitespace().map(|s| s.to_string()).collect::<Vec<String>>();
    let mut hyp_tokens = hypothesis.to_lowercase().split_whitespace().map(|s| s.to_string()).collect::<Vec<String>>();
    if (!hyp_tokens || !ref_tokens) {
        0.0_f64
    }
    let mut ref_set = ref_tokens.into_iter().collect::<HashSet<_>>();
    let mut matches = hyp_tokens.iter().filter(|t| ref_set.contains(&t)).map(|t| 1).collect::<Vec<_>>().iter().sum::<i64>();
    (matches / hyp_tokens.len())
}

/// ROUGE-L F1 via LCS (pure Python).
pub fn _rouge_l(reference: String, hypothesis: String) -> f64 {
    // ROUGE-L F1 via LCS (pure Python).
    let mut r#ref = reference.to_lowercase().split_whitespace().map(|s| s.to_string()).collect::<Vec<String>>();
    let mut hyp = hypothesis.to_lowercase().split_whitespace().map(|s| s.to_string()).collect::<Vec<String>>();
    if (!r#ref || !hyp) {
        0.0_f64
    }
    let (mut m, mut n) = (r#ref.len(), hyp.len());
    let mut dp = 0..(m + 1).iter().map(|_| (vec![0] * (n + 1))).collect::<Vec<_>>();
    for i in 1..(m + 1).iter() {
        for j in 1..(n + 1).iter() {
            dp[&i][j] = if r#ref[(i - 1)] == hyp[(j - 1)] { (dp[(i - 1)][(j - 1)] + 1) } else { dp[(i - 1)][&j].max(dp[&i][(j - 1)]) };
        }
    }
    let mut lcs = dp[&m][&n];
    let mut precision = (lcs / n);
    let mut recall = (lcs / m);
    if (precision + recall) == 0 {
        0.0_f64
    }
    (((2 * precision) * recall) / (precision + recall))
}
